### DISCUSSION TITLE
NOCRAM  availability AND Mapped data for air temperature (BLUEWS)

### DISCUSSION THREAD

**Discussion by HibaHamdi09** [UMEP-dev/UMEP#326] [Q&A] [answered]
NOCRAM  availability: In the paper UMEP - An integrated tool for urban climatology and climate-sensitive planning applications ( Lindberg 2015), the module NOCRAM is mentioned. In the context of my Ph.D on urban heat islands, I need to produce night air temperature maps. Could you tell me if this module will be available soon as a plug-in and/or as a source code?

Mapped data for air temperature (BLUEWS) : After searching in the literature and manuals, I only found validation plots of predicted air temperature against ground truth values. Is it possible to generate a map of air temperature from BLUEWS outputs?

Thank you in advance for your  help.

**Comment by biglimp** [comment, marked as answer]
I am in the stage of incorporating the urban weather generator (https://github.com/ladybug-tools/uwg) into UMEP. That will produce the output you are looking for. Maybe you are interested in collaborating in this work? Are you familiar with Python?

**Reply by HibaHamdi09** [reply]
Thank you for your reply.
If I understood correctly, the UWG would allow me to generate day and night air temperature maps?
Is it already possible to do this with the python codes provided on github?  Or shall I wait for the incorporation in UMEP (QGIS)?

I mostly work with remote sensing data and I’m looking to produce air temperature maps for calibration and validation purposes. I have a basic level in Python and I’m not very familiar with your model so, for now, I’m not sure how I could help you. But If I develop something that may be useful for you, I will be happy to share it.

And of course, your model will be cited in case of publication!

**Reply by biglimp** [reply]
You could use UWG as it is. The incorporation in UMEP will include data preparation and post processing for UWG, directly in QGIS.

**Reply by biglimp** [reply]
I am doing it inbetween other things so I could not tell you an exact date but hopefully before the end of this year.

**Reply by HibaHamdi09** [reply]
thank you for your reactivity! And good luck

**Comment by HibaHamdi09** [comment]
Great! Could you tell me please when the incorporation in UMEP would be finished?

**Reply by biglimp** [reply]
Maybe in a couple of months

**Comment by rarygit** [comment]
@biglimp 

If it helps, just one observation regarding incorporating the UWG model.

These peer reviews were rather scathing of the UWG rural model assumptions (Bueno et al) when incorporated into another model:
https://gmd.copernicus.org/articles/14/961/2021/gmd-14-961-2021-discussion.html
https://gmd.copernicus.org/preprints/gmd-2019-176/gmd-2019-176-RC1.pdf
https://gmd.copernicus.org/preprints/gmd-2019-176/gmd-2019-176-RC2.pdf

The VCWG model was subsequently revised to address these issues
https://gmd.copernicus.org/preprints/gmd-2019-176/gmd-2019-176-AC1-supplement.pdf
"In brief, we have replaced the rural model, which originally had too many unjustified constants, with a well-established surface layer model, i.e. the Monin-Obukhov Similarity Theory (MOST). This fundamental change has resulted in improved performance of the VCWG model. "

**Reply by biglimp** [reply]
Thanks. Yes, I am aware of this. The reason of taking UWG is its translation to Python3 makes it easy to incorporate. Maybe VGWG also comes in a Python version? Do you know anything about this @rarygit?

**Reply by rarygit** [reply]
Hi @biglimp!

I successfully ran the VCWG v1.3.2.py simulation under python v3.9.2, using a conda environment
(conda create -n myenv python=3.9.2).
And I had to modify the Utilities.py in VCWG by looking at the uwg utilities.py file (ladybug version)
Perhaps because the ladybug version was already python 3... maybe, can't remember exactly

Previously I ran UWG in the QGIS python console using pypi as the installer in the conda env.

I converted VCWG v1.32 on GitHub (python 2.7) to python 3 using python module "2to3"
https://docs.python.org/3/library/2to3.html

Conversion was straightforward: 

Every time I ran VCWG at the command line, it would report an error identifying which python 2 script needed converting.
Then I ran 2to3 over that particular script using the same command line, then reran VCWG. 
Next script came up as error and was then converted, rerun, next script ... in a step-by-step process.
Somewhat crude, but it worked for me as a novice doing this conversion for the first time.

Eventually I had a working version of VCWG in a python 3 conda environment.
I sent a copy of that to the authors at the beginning of March this year.
I have attached it here below:

[Attachments-[VCWG v1.3.2] Python 3 version.zip](https://github.com/UMEP-dev/UMEP/files/7498491/Attachments-.VCWG.v1.3.2.Python.3.version.zip)

Prof. Amir Aliabiadi has now opened his own repository and released VCWG v1.4.4
https://github.com/AmirAAliabadi/VCWGv1.4.4
This uses python v2.18, which I tested OK on Ubuntu 20.04 
I have not had time to convert v1.4.4 to python 3.
I had rather hoped that Amir would have started development in python 3.

Open access copy of publication for VCWG v1.4.4 is here:
https://www.researchgate.net/publication/353042678_How_Much_Building_Renewable_Energy_Is_Enough_The_Vertical_City_Weather_Generator_VCWG_v144

I asked Amir a question regarding the use of VCWG, which he answered
https://github.com/AmirAAliabadi/VCWGv1.4.4/issues/1

I also made the comment that "In the User Manual, to retrieve ERA5 climate data in epw and csv formats, it might be helpful to suggest using the website www.shinyweatherdata.com", which he did.

**Reply by biglimp** [reply]
So did you try your VCWG code directly in QGIS, without conda? This is needed to incorporate in UMEP. We need to have something stable that works on most OS and for regular users. Hopefully a translation to P3 (similar to UWG) will come, then we can shift from UWG to VCWG in UMEP

**Reply by rarygit** [reply]
Directly tested OK early this morning in QGIS v3.16 python console/editor, Ubuntu 20.04, python v3.9.2

Should run on all OSes with this modification to the main python script, VCWGv132.py
i.e. add the path to the folder which contains sub-folder UWG

For example, in my case the path is:

import sys
sys.path.insert(0, '/media/l1nux/BAK3_SEA4TB/ALL_Unsorted/VCWG132/VCWGv132/')
from UWG import UWG

[VCWGv132_py_only.zip](https://github.com/UMEP-dev/UMEP/files/7517873/VCWGv132_py_only.zip)

As stated in the VCWG manual, don't forget to discard output for the first 24 hours.

I believe it will be up to us to make the translation of VCWG to python 3.
Using "2to3" has worked in relation to "accepted" syntax changes.
Without any fiurther optimisation of numpy 
(using Geany to quick search for "numpy" in files in UWG sub-folder: DragTurb.py, ColModel.py, RadiationFunctions.py, NumericalSolver.py etc) 

You could correspond with Amir directly on his GitHub page to see if his team (Uni Guelph) can work directly within Python 3.
Especially to facilitate its incoporation into UMEP. I think this would be persuasive.

**Reply by biglimp** [reply]
Did you not submit your P3 translation to Amir? As he is maintainer of VCWG, he should be the one who has the code since he might update the model etc.

**Reply by rarygit** [reply]
Yes, I did

I just checked Amir's GitHub repo and 12 hours ago he updated VCWG to v1.4.5
The documentation states that it requires Python 2.7.13 amd64
https://github.com/AmirAAliabadi/VCWGv1.4.5

I have posted an issue there and ask Amir if he would like to join our discussion here, regarding UMEP and the development of VCWG in P3.

https://github.com/AmirAAliabadi/VCWGv1.4.5/issues/1

Notwithstanding that option, I took the view that we are taking the initiative as VCWG is a better option for UMEP than UWG.
If it comes to that, you/I could compare v1.4.5 with 1.3.2 and see which files were changed.
That would mean reducing the number of v1.4.5 scripts to convert to P3

**Reply by AmirAAliabadi** [reply]
Dear All,

Thank you for considering VCWG and trying v1.3.2 and v1.4.5. Both versions are essentially the same with v1.4.5 including renewable energy options. Both versions are designed to run on Python 2.7. We currently do not have the resources to translate these versions from Python 2.7 to 3. So you may wish to try to translate them with your own effort. Please note that the publications pertaining to these versions are available here:

Moradi, M., Dyer, B., Nazem, A., Nambiar, M. K., Nahian, M. R., Bueno, B., Mackey, C., Vasanthakumar, S., Nazarian, N., Krayenhoff, E. S., Norford, L. K., & Aliabadi, A. A. (2021), 'The Vertical City Weather Generator (VCWG v1.3.2)', Geoscientific Model Development, 14(2), 961-984, doi:10.5194/gmd-14-961-2021.

Aliabadi, A. A., Moradi, M., McLeod, R. M., Calder, D., & Dernovsek, R. (2021), 'How Much Building Renewable Energy Is Enough? The Vertical City Weather Generator (VCWG v1.4.4)', Atmosphere, 12(7), 882, doi:10.3390/atmos12070882.

The good news is that we have yet developed a new version of **VCWG v2.0.0** which completely runs on Python 3. This is an improved model which **1) offers hydrological modelling**, and **2) in addition to MOST, it offers the ability to be forced from the top of the urban domain using mesoscale data products (e.g. ERA-5 pressure levels)**. In fact, with this option the rural model can be bypassed in this version. However, due to more physical modelling, this version has greater number of input variables (soil parameters, vegetation parameters, etc.). You can read more about this version here:

Moradi, M., Krayenhoff E. S., & Aliabadi, A. A. (2022), 'A comprehensive indoor-outdoor urban climate model with hydrology: The Vertical City Weather Generator (VCWG v2.0.0)', Building and Environment, 207(B), 108406, doi:10.1016/j.buildenv.2021.108406.

Moradi, M. (2021), Ph.D., 'The Vertical City Weather Generator', University of Guelph, Guelph, Canada.

We do not intend to publish VCWG v2.0.0 on GitHub yet, but we will be happy to provide it to you privately on dropbox (etc.) You can include it in UMEP. The license is GNU General Public License v3.0. We like to continue this conversation via **email**. Please contact me at **aliabadi@uoguelph.ca** and introduce **yourselves** and **affiliations**.

Regards,
Amir A. Aliabadi

**Reply by rarygit** [reply]
@AmirAAliabadi 
Many thanks for the quick reply and the lucid description of VCWG v2, running on Python 3. 
Your article (doi:10.1016/j.buildenv.2021.108406) comprehensively presents this latest model.  
I have a few questions regarding the subcanopy convective fluxes and urban pipe drainage, so I will continue this conversation via email.

@biglimp 
I hope that VCWG v2 can now be integrated into UMEP.
I am happy to help with the testing.

**Comment by biglimp** [comment]
Thanks Amir. We will wait to include VGWG v2 in UMEP until it is properly released in a Github repo.
